storage/raft_log_queue.go

```diff
@@ -0,0 +1,127 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Bram Gruneir (bram+code@cockroachlabs.com)
+
+package storage
+
+import (
+	"time"
+
+	"github.com/cockroachdb/cockroach/client"
+	"github.com/cockroachdb/cockroach/config"
+	"github.com/cockroachdb/cockroach/gossip"
+	"github.com/cockroachdb/cockroach/roachpb"
+	"github.com/cockroachdb/cockroach/util"
+	"github.com/cockroachdb/cockroach/util/log"
+)
+
+const (
+	// raftLogQueueMaxSize is the max size of the queue.
+	raftLogQueueMaxSize = 100
+	// RaftLogQueueTimerDuration is the duration between checking the
+	// raft logs.
+	RaftLogQueueTimerDuration = time.Second
+	// RaftLogQueueLogSizeThreshold is the number of raft log entries after
+	// which truncation is required.
+	RaftLogQueueLogSizeThreshold = 1
+)
+
+// raftLogQueue manages a queue of replicas slated to have their raft logs
+// truncated by removing unneeded entries.
+type raftLogQueue struct {
+	baseQueue
+	db *client.DB
+}
+
+// newRaftLogQueue returns a new instance of raftLogQueue.
+func newRaftLogQueue(db *client.DB, gossip *gossip.Gossip) *raftLogQueue {
+	rlq := &raftLogQueue{
+		db: db,
+	}
+	rlq.baseQueue = makeBaseQueue("raftlog", rlq, gossip, raftLogQueueMaxSize)
+	return rlq
+}
+
+func (*raftLogQueue) needsLeaderLease() bool {
+	return true
+}
+
+func (*raftLogQueue) acceptsUnsplitRanges() bool {
+	return true
+}
+
+// getTruncatableIndexes looks up the replica's raft status and returns the
+// current first index for the replica and the oldest still in use index.
+func getTruncatableIndexes(r *Replica) (uint64, uint64, error) {
+	desc := r.Desc()
+	raftStatus := r.store.RaftStatus(desc.RangeID)
+	if raftStatus == nil {
+		return 0, 0, util.Errorf("could not get raft status for %s", desc)
+	}
+	firstIndex, err := r.FirstIndex()
```

Comment:
Move this to the beginning or end of the function instead of in the middle of the stuff using `raftStatus`. (or just call `FirstIndex` from `shouldQueue` and `process`; putting it in this function is awkward)